specs/crux/beacon-chain.md

```diff
@@ -0,0 +1,247 @@
+        withdrawable_epoch=FAR_FUTURE_EPOCH,
+        effective_balance=effective_balance,
+    )
+```
+
+##### `process_delegation`
+
+```python
+def process_delegation(state: BeaconState, delegation: Delegation) -> None:
+    message = delegation.message
+    current_epoch = get_current_epoch(state)
+    assert  message.epoch <= current.epoch <= message.epoch + MAX_DELEGATION_INCLUSION_DELAY
+
+    delegating_index = message.delegating_index
+    min_delegating_balance = MAX_EFFECTIVE_BALANCE + MIN_DEPOSIT_AMOUNT + DELEGATION_TRANSACTION_COST
+    assert state.balances[delegating_index] >= min_delegating_balance
+    amount = state.balances[delegating_index] - MAX_EFFECTIVE_BALANCE - DELEGATION_TRANSACTION_COST
+
+    delegating_validator = state.validators[delegating_index]
+    assert is_active_validator(delegating_validator, current_epoch)
+    assert not delegating_validator.slashed 
+
+    assert delegating_validator.delegate == delegating_index
+
+    delegating_pubkey = message.delegating_pubkey
+    assert hash(delegating_pubkey)[1:] == delegating_validator.withdrawal_credentials[1:]
+
+    domain = compute_domain(DOMAIN_DELEGATION)
+    signing_root = compute_signing_root(message, domain)
+    assert bls.Verify(delegating_pubkey, signing_root, delegation.signature)
+
+    validator_pubkeys = [v.pubkey for v in state.validators]
+    if pubkey not in validator_pubkeys:
+        index = len(state.validators)
+        state.validators.append(get_validator_from_delegation(message, amount))
+    else:
+        index = ValidatorIndex(validator_pubkeys.index(pubkey))
+        assert index > delegating_index
+        assert state.validators[index].exit_epoch == FAR_FUTURE_EPOCH
+
+    delegating_validator.delegate = index
+    proposer_index = get_beacon_proposer_index(state)
+    decrease_balance(state, delegating_validator, amount + DELEGATION_TRANSACTION_COST)
+    increase_balance(state, index, amount)
+    increase_balance(state, proposer_index, DELEGATION_TRANSACTION_COST)
+```
+
+### Epoch processing
+
+```python
+def process_epoch(state: BeaconState) -> None:
+    process_justification_and_finalization(state)
+    process_inactivity_updates(state)
+    process_rewards_and_penalties(state)
+    process_delegation_transfers(state)  # [New in Crux]
+    process_registry_updates(state)   # [Modified in Crux]
+    process_slashings(state)
+    process_eth1_data_reset(state)
+    process_effective_balance_updates(state)
+    process_slashings_reset(state)
+    process_randao_mixes_reset(state)
+    process_historical_roots_update(state)
+    process_participation_flag_updates(state)
+    process_sync_committee_updates(state)
+```
+
+#### `process_delegation_transfers`
+
+```python
+def process_delegation_transfers(state: BeaconState) -> None:
+    delegating_validators = [(i, v) for i, v in enumerate(state.validators) if v.delegate > i and not v.slashed]
+    for index, validator in delegating_validators:
```

Comment:
So the extra load for validators for delegation is a full loop on the entire validator set + potentially modify the entire balances array. Is it really necessary to be done every epoch? Maybe every N to reduce overall load of the feature

Reply:
In principle this can be done within `process_rewards_and_penalties` avoiding any extra loop and any extra load costs. It is simply a check before applying a reward if it has to go to the delegate or not. I chose the current approach because of its simplicity to gauge interest in this type of application or not. You are right however that this can be done every `N` epochs.

Reply:
Definitely agree this is a very useful feature to reduce load in the queues :+1:

Reply:
Tho in other POS networks delegate is a widely use term that needs something different from what this PR achieves. Would you be a better term for the action of moving excess balance to another validator?

Reply:
Thanks, surely delegation will not be the final term, every single person that reacted positively to this PR pointed that this term is commonly used in a different way, if this gets any traction hopefully someone will come up with the right term. FWIW, an implementation that doesn't have an extra loop is 
```python
def process_rewards_and_penalties(state: BeaconState) -> None:
    # No rewards are applied at the end of `GENESIS_EPOCH` because rewards are for work done in the previous epoch
    if get_current_epoch(state) == GENESIS_EPOCH:
        return

    flag_deltas = [get_flag_index_deltas(state, flag_index) for flag_index in range(len(PARTICIPATION_FLAG_WEIGHTS))]
    deltas = flag_deltas + [get_inactivity_penalty_deltas(state)]
    for (rewards, penalties) in deltas:
        for index, val in enumerate(state.validators):
            decrease_balance(state, ValidatorIndex(index), penalties[index])
            if rewards[index] > 0:
                new_balance = state.balances[index] + rewards[index]
                to_delegate = false
                if val.delegate > index and not val.slashed and new_balance > MAX_EFFECTIVE_BALANCE:
                    delegate_validator = state.validators[val.delegate]
                    if delegate_validator.exit_epoch == FAR_FUTURE_EPOCH:
                        to_delegate = true
                if to_delegate: 
                    state.balance[index] = MAX_EFECTIVE_BALANCE
                    increase_balance(state, val.delegate, new_balance - MAX_EFFECTIVE_BALANCE)
                else:
                    state.balance[index] = new_balance
```
But I think actually the extra loop only over the delegating indices is more efficient in compiled languages. And regardless of any choice, implementers can do either one cause it does not affect the final state transition. The biggest cost of this PR I think is in signature verification and block size, not here. But at `MAX_DELEGATIONS = 16`  in only 1 day blocks should be back to empty. And we currently process 16 signatures without any problem.